src/GRA.Data/Repository/AvatarBundleRepository.cs

```diff
@@ -21,6 +21,47 @@ public class AvatarBundleRepository
         {
         }
 
+        public async Task<List<UserLog>> UserHistoryAsync(int userId)
+        {
+            var userLogs = await _context.UserLogs
+               .AsNoTracking()
+               .Where(_ => _.UserId == userId
+                      && !_.IsDeleted)
+               .OrderBy(_ => _.CreatedAt)
+               .ProjectTo<UserLog>(_mapper.ConfigurationProvider)
+               .ToListAsync();
+
+            foreach (var userLog in userLogs)
```

Comment:
You can combine the `foreach` and the `if` using Linq: `foreach (var userLog in userLogs.Where(_ => _.BadgeId != null)`